File: packages/linux-dashboard/src/main/launcherWindow.ts
import { BrowserWindow } from 'electron';
import { buildUrl } from './appProtocol';
import { registerNativeMessages } from './nativeMessages';
import { bridgePreloadPath } from './panelWindow';
import { configureWebContents } from './webContentsSetup';

export class LauncherWindow {
  private window: BrowserWindow | null = null;
  private showWhenReady = false;
  private readonly appVersion: string;
  private readonly onAddWidget: (widgetType: number) => void;

  constructor(appVersion: string, onAddWidget: (widgetType: number) => void) {
    this.appVersion = appVersion;
    this.onAddWidget = onAddWidget;
  }

  show(): void {
    if (this.window && !this.window.isDestroyed()) {
      if (this.window.isMinimized()) this.window.restore();
      this.window.show();
      this.window.focus();
      return;
    }

    const win = new BrowserWindow({
      show: false,
      width: 850,
      height: 580,
      minWidth: 700,
      minHeight: 540,
      title: 'Add Widget — Classroom Widgets',
      autoHideMenuBar: true,
      backgroundColor: '#F5F3F1',
      webPreferences: {
        preload: bridgePreloadPath(),
        contextIsolation: true,
        sandbox: false,
        nodeIntegration: false,
        additionalArguments: [`--classroom-app-version=${this.appVersion}`],
      },
    });
    this.window = win;
    this.showWhenReady = true;
    configureWebContents(win.webContents);
    registerNativeMessages(win.webContents, (message) => this.handleMessage(message));
    win.once('ready-to-show', () => {
      if (!this.showWhenReady || win.isDestroyed()) return;
      win.show();
      win.focus();
    });
    win.once('closed', () => {
      this.showWhenReady = false;
      if (this.window === win) this.window = null;
    });
    void win.loadURL(buildUrl({ surface: 'widget-launcher' }));
  }

  close(): void {
    this.showWhenReady = false;
    this.window?.close();
  }

  private handleMessage(message: Record<string, unknown>): void {
    if (message.handler !== 'classroomDashboard' || message.schemaVersion !== 1) return;
    if (message.type === 'desktop-launcher-close') {
      this.close();
      return;
    }
    if (message.type === 'desktop-launcher-add-widget'
      && typeof message.widgetType === 'number' && Number.isInteger(message.widgetType)) {
      this.onAddWidget(message.widgetType);
    }
  }
}

File: packages/windows-dashboard/Source/LauncherWindow.cs
using System.ComponentModel;
using System.Text.Json;
using System.Windows;
using Microsoft.Web.WebView2.Core;
using Microsoft.Web.WebView2.Wpf;

namespace ClassroomWidgets;

public sealed class LauncherWindow
{
    private readonly Window _window;
    private readonly WebView2 _webView;
    private readonly Action<int> _addWidget;
    private bool _initializing;
    private bool _ready;
    private bool _showWhenReady;

    public LauncherWindow(Action<int> addWidget)
    {
        _addWidget = addWidget;
        _webView = new WebView2();
        _window = new Window
        {
            Title = "Add Widget — Classroom Widgets",
            Width = 850,
            Height = 580,
            MinWidth = 700,
            MinHeight = 540,
            WindowStartupLocation = WindowStartupLocation.CenterScreen,
            Background = System.Windows.Media.Brushes.White,
            Content = _webView
        };
        _window.Closing += OnClosing;
    }

    public void Show()
    {
        _showWhenReady = true;
        if (_ready)
        {
            ShowReadyWindow();
            return;
        }
        if (!_initializing) _ = InitializeAsync();
    }

    public void Hide()
    {
        _showWhenReady = false;
        _window.Hide();
    }

    private async Task InitializeAsync()
    {
        _initializing = true;
        try
        {
            await DashboardWebView.InitializeAsync(_webView, string.Empty);
            var core = _webView.CoreWebView2;
            core.WebMessageReceived += OnWebMessageReceived;
            core.NavigationCompleted += (_, args) =>
            {
                if (!args.IsSuccess)
                {
                    DashboardLog.Error($"Widget launcher navigation failed: {args.WebErrorStatus}");
                    return;
                }
                _ready = true;
                if (_showWhenReady) ShowReadyWindow();
            };
            core.Navigate(DashboardWebView.BuildUrl(new Dictionary<string, string>
            {
                ["surface"] = "widget-launcher"
            }));
        }
        catch (Exception error) when (error is WebView2RuntimeNotFoundException or System.Runtime.InteropServices.COMException)
        {
            DashboardLog.Error($"Unable to open widget launcher: {error.Message}");
        }
        finally
        {
            _initializing = false;
        }
    }

    private void ShowReadyWindow()
    {
        if (_window.WindowState == WindowState.Minimized) _window.WindowState = WindowState.Normal;
        _window.Show();
        _window.Activate();
    }

    private void OnClosing(object? sender, CancelEventArgs args)
    {
        if (App.IsShuttingDown) return;
        args.Cancel = true;
        Hide();
    }

    private void OnWebMessageReceived(object? sender, CoreWebView2WebMessageReceivedEventArgs args)
    {
        if (!DashboardWebView.IsAllowed(args.Source)) return;
        JsonDocument document;
        try
        {
            document = JsonDocument.Parse(args.WebMessageAsJson);
        }
        catch (JsonException)
        {
            return;
        }

        using (document)
        {
            var body = document.RootElement;
            if (body.ValueKind != JsonValueKind.Object
                || !body.TryGetProperty("handler", out var handler) || handler.GetString() != "classroomDashboard"
                || !body.TryGetProperty("schemaVersion", out var schemaVersion) || !schemaVersion.TryGetInt32(out var version) || version != 1
                || !body.TryGetProperty("type", out var type)) return;

            if (type.GetString() == "desktop-launcher-close")
            {
                Hide();
                return;
            }
            if (type.GetString() == "desktop-launcher-add-widget"
                && body.TryGetProperty("widgetType", out var widgetType)
                && widgetType.TryGetInt32(out var value))
            {
                _addWidget(value);
            }
        }
    }
}
